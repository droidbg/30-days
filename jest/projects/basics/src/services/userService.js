// Used for Test of Mocking an API Call

async function fetchUsers() {
    const res = await fetch("https://jsonplaceholder.typicode.com/users");
    if (!res.ok) throw new Error("Failed to fetch users");
    return res.json();
  }
  module.exports = { fetchUsers };
  