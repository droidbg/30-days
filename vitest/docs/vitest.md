# Vitest
Vitest = a testing framework for JavaScript/TypeScript.

Built to work well with Vite + modern apps (React, Vue, Node).

- Runs unit tests (small pieces of code)

- Snapshot testing (check UI looks the same)

- Mocks (fake versions of APIs or services)

### - Node.js Project

```bash
npm init -y
npm install -D vitest
```

``` javascript
package.json
"scripts": {
  "test": "vitest",
  "test:run": "vitest run",
  "coverage": "vitest run --coverage"
}

```


### - React Project

```bash
npm create vite@latest vitest-react -- --template react-ts
cd vitest-react
npm install
npm install -D vitest @testing-library/react @testing-library/jest-dom jsdom
```

In `vite.config.ts` , enable testing

```ts
/// <reference types="vitest" />
import { defineConfig } from 'vitest/config'
import react from "@vitejs/plugin-react";

export default defineConfig({
  plugins: [react()],
  test: {
    environment: "jsdom", // simulate browser
    globals: true,
    setupFiles: "./src/setupTests.ts",
  },
});

```


### - Vue Project
```bash
npm create vite@latest vitest-vue -- --template vue-ts
cd vitest-vue
npm install
npm install -D vitest @vue/test-utils jsdom
```
In `vite.config.ts` , enable testing

```ts
/// <reference types="vitest" />
import { defineConfig } from "vite";
import vue from "@vitejs/plugin-vue";

export default defineConfig({
  plugins: [vue()],
  test: {
    environment: "jsdom",
    globals: true,
  },
});
```

 

## Snapshot Test
Snapshot = save HTML structure and compare later.
```tsx
import { render } from "@testing-library/react";
import { test, expect } from "vitest";

function Button() {
  return <button>Click Me</button>;
}

test("button snapshot", () => {
  const { container } = render(<Button />);
  expect(container).toMatchSnapshot();
});
```
- `render(<Button />)` → renders React component.

- `expect(container).toMatchSnapshot();` → saves output as snapshot.
- First run creates `__snapshots__` folder, later runs compare automatically.
- Next run → Vitest checks if UI changed.


#### Coverage Report

```bash
npm install -D c8
npm run coverage
npx vitest run --coverage
```