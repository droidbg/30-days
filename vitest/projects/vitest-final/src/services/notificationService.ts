export async function sendEmail(to: string, message: string) {
  // Imagine this calls SendGrid/Mailgun
  console.log(`Email sent to ${to}: ${message}`);
  return { success: true };
}
