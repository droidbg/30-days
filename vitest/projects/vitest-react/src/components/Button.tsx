type Props = { label: string; onClick?: () => void; disabled?: boolean };


export function Button({ label, onClick, disabled }: Props) {
  return (
    <button disabled={disabled} onClick={onClick}>
      {label}
    </button>
  );
}
