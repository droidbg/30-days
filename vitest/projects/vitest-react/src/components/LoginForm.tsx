import { useState } from "react";

export function LoginForm({ onLogin }: { onLogin: (u: string, p: string) => void }) {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");

  return (
    <form onSubmit={e => { e.preventDefault(); onLogin(username, password); }}>
      <input
        placeholder="Username"
        value={username}
        onChange={e => setUsername(e.target.value)}
      />
      <input
        type="password"
        placeholder="Password"
        value={password}
        onChange={e => setPassword(e.target.value)}
      />
      <button type="submit">Login</button>
    </form>
  );
}
