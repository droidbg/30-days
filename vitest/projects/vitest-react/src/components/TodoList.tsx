import { useState } from "react";

export function TodoList() {
  const [todos, setTodos] = useState<string[]>([]);

  function addTodo(todo: string) {
    setTodos([...todos, todo]);
  }

  return (
    <div>
      <button onClick={() => addTodo("New Task")}>Add Todo</button>
      <ul>
        {todos.map((t, i) => (
          <li key={i}>{t}</li>
        ))}
      </ul>
    </div>
  );
}
