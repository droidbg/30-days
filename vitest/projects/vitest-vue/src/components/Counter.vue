<template>
  <div>
    <p>Count: {{ count }}</p>
    <button @click="count++">Increment</button>
  </div>
</template>

<script setup>
import { ref } from "vue";
const count = ref(0);
</script>