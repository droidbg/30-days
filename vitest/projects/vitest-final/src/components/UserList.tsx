import { useEffect, useState } from "react";
import { getUsers } from "../services/userService";

export function UserList() {
  const [users, setUsers] = useState<any[]>([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    getUsers().then(data => {
      setUsers(data);
      setLoading(false);
    });
  }, []);

  if (loading) return <p>Loading...</p>;

  return (
    <ul>
      {users.map((u) => (
        <li key={u.id}>{u.name}</li>
      ))}
    </ul>
  );
}
