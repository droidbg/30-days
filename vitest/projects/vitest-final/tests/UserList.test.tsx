import { render, screen, waitFor } from "@testing-library/react";
import { test, expect, vi } from "vitest";
import { UserList } from "../src/components/UserList";
import * as userService from "../src/services/userService";

test("renders users from API", async () => {
  vi.spyOn(userService, "getUsers")
    .mockResolvedValue([{ id: 1, name: "Alice" }, { id: 2, name: "Bob" }]);

  render(<UserList />);

  expect(screen.getByText("Loading...")).toBeInTheDocument();

  await waitFor(() => {
    expect(screen.getByText("Alice")).toBeInTheDocument();
    expect(screen.getByText("Bob")).toBeInTheDocument();
  });
});

test("matches snapshot", async () => {
  vi.spyOn(userService, "getUsers")
    .mockResolvedValue([{ id: 1, name: "Alice" }]);

  const { asFragment } = render(<UserList />);
  await screen.findByText("Alice");

  expect(asFragment()).toMatchSnapshot();
});
