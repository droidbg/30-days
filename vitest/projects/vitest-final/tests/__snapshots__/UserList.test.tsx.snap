// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <ul>
    <li>
      Alice
    </li>
  </ul>
</DocumentFragment>
`;
